controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,531 @@
+	ec.log.Errorf("dropped Service %s out of update queue: %v", key, err)
+}
+
+// === Callbacks ===
+
+// When a service update has been received (regardless of the event type, i.e.
+// can be Added, Modified, Deleted) send it to the endpoint controller for
+// processing.
+func (ec *EndpointsController) updateService(obj interface{}) {
+	svc, ok := obj.(*corev1.Service)
+	if !ok {
+		ec.log.Errorf("error processing Service event: expected *corev1.Service, got %#v", obj)
+		return
+	}
+
+	if svc.Namespace == "kube-system" {
+		ec.log.Tracef("skipping Service event: %s/%s is a kube-system Service", svc.Namespace, svc.Name)
+		return
+	}
+	// Use client-go's generic key function to make a key of the format
+	// <namespace>/<name>
+	key, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		ec.log.Infof("failed to get key for svc %s/%s: %v", svc.Namespace, svc.Name, err)
+	}
+
+	ec.queue.Add(key)
+}
+
+func isReady(ew *ewv1alpha1.ExternalWorkload) bool {
+	if len(ew.Status.Conditions) == 0 {
+		return false
+	}
+
+	// Loop through the conditions and look at each condition in turn starting
+	// from the top.
+	for i := range ew.Status.Conditions {
+		cond := ew.Status.Conditions[i]
+		// Stop once we find a 'Ready' condition. We expect a resource to only
+		// have one 'Ready' type condition.
+		if cond.Type == ewv1alpha1.WorkloadReady && cond.Status == ewv1alpha1.ConditionTrue {
+			return true
+		}
+	}
+
+	return false
+}
+
+// === Util ===
+
+// Check whether two label sets are matching
+func labelsChanged(old, updated *ewv1alpha1.ExternalWorkload) bool {
+	if len(old.Labels) != len(updated.Labels) {
+		return true
+	}
+
+	for upK, upV := range updated.Labels {
+		oldV, ok := old.Labels[upK]
+		if !ok || oldV != upV {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Check whether two workload resources have changed
+// Note: we are interested in changes to the ports, ips and readiness fields
+// since these are going to influence a change in a service's underlying
+// endpoint slice
+func workloadChanged(old, updated *ewv1alpha1.ExternalWorkload) bool {
+	if isReady(old) != isReady(updated) {
+		return true
+	}
+
+	ports := make(map[int32]struct{})
+	for _, pSpec := range updated.Spec.Ports {
+		ports[pSpec.Port] = struct{}{}
+	}
+
+	for _, pSpec := range old.Spec.Ports {
+		if _, ok := ports[pSpec.Port]; !ok {
+			return true
+		}
+	}
+
+	ips := make(map[string]struct{})
+	for _, addr := range updated.Spec.WorkloadIPs {
+		ips[addr.Ip] = struct{}{}
+	}
+	for _, addr := range old.Spec.WorkloadIPs {
+		if _, ok := ips[addr.Ip]; !ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+// servicesToUpdate accepts pointers to two ExternalWorkload resources used to
+// determine the state of an update. Based on the state of the update and the
+// references to the workload resources, it will collect a set of Services that
+// need to be updated.
+func (ec *EndpointsController) servicesToUpdate(old, updated *ewv1alpha1.ExternalWorkload) []string {
```

Comment:
I'm having  a really hard time reasoning through this method and figuring out if it's correct.  lots of different cases and hard to keep track of them all.  can it written in an easier to understand way?

if labels have changed, then membership has changed and we need to recompute services for both the old and new workload
otherwise if only workload has changed then membership has not changed and services for old and new workload should be the same

so doesn't this all amount to the same thing? as long as we're dedupping, shouldn't we always recompute services for both old and new?

Reply:
@adleong @mateiidavid  I am having a bit of a hard time following myself. This is why I tried to rewrite it in a different way and added some comments. Let me all know whether this is correct and whether it looks simpler. Also. please make sure to write tests of it that exercise all the edge cases. 

```golang
func toSet(s []string) map[string]struct{} {
	set := map[string]struct{}{}
	for _, k := range s {
		set[k] = struct{}{}
	}
	return set
}

// servicesToUpdate accepts pointers to two ExternalWorkload resources used to
// determine the state of an update. Based on the state of the update and the
// references to the workload resources, it will collect a set of Services that
// need to be updated.
func (ec *EndpointsController) servicesToUpdate(old, updated *ewv1alpha1.ExternalWorkload) []string {
	labelsChanged := labelsChanged(old, updated)
	workloadChanged := workloadChanged(old, updated)
	if !labelsChanged && !workloadChanged {
		ec.log.Debugf("skipping ExternalWorkload update; nothing has changed between old rv %s and new rv %s", old.ResourceVersion, updated.ResourceVersion)
		return nil
	}

	newSelection, err := ec.getSvcMembership(updated)
	if err != nil {
		ec.log.Errorf("failed to get service membership for workload %s/%s: %v", updated.Namespace, updated.Name, err)
		return nil
	}

	oldSelection, err := ec.getSvcMembership(updated)
	if err != nil {
		ec.log.Errorf("failed to get service membership for workload %s/%s: %v", updated.Namespace, updated.Name, err)
		return nil
	}

	result := map[string]struct{}{}
	if labelsChanged && workloadChanged {
		// we need to consider the union of the selection
		// as both selection and workload content
		// has changed
		for _, s := range append(newSelection, oldSelection...) {
			result[s] = struct{}{}
		}
	} else if workloadChanged {
		// workload changed, but selection stays the same
		// in this case newSelection and oldSelection should
		// be the same, so we can safely reconsider all services
		// from newSelection
		return newSelection
	} else {
		// only selection changed, so we need to consider the diff
		// of the new and old selection as the rest of the services
		// should stay the same as content of workload has not changed
		newSelectionSet := toSet(newSelection)
		oldSelectionSet := toSet(oldSelection)

		// consider all services that are in the
		// old selection, but are not in the new
		// one (i.e. removed ones)
		for _, oldSvc := range oldSelection {
			if _, ok := newSelectionSet[oldSvc]; !ok {
				result[oldSvc] = struct{}{}
			}
		}

		// consider all services that are in the
		// new selection, but are not in the old
		// one (i.e. added ones)
		for _, newSvc := range newSelection {
			if _, ok := oldSelectionSet[newSvc]; !ok {
				result[newSvc] = struct{}{}
			}
		}
	}

	var resultSlice []string
	for svc, _ := range result {
		resultSlice = append(resultSlice, svc)
	}
	return resultSlice
}
```